src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -958,26 +963,23 @@ void PSParallelCompact::summary_phase(bool maximum_compaction)
 // may be true because this method can be called without intervening
 // activity.  For example when the heap space is tight and full measure
 // are being taken to free space.
-bool PSParallelCompact::invoke(bool maximum_heap_compaction) {
+bool PSParallelCompact::invoke(bool clear_all_soft_refs) {
   assert(SafepointSynchronize::is_at_safepoint(), "should be at safepoint");
   assert(Thread::current() == (Thread*)VMThread::vm_thread(),
          "should be in vm thread");
 
-  ParallelScavengeHeap* heap = ParallelScavengeHeap::heap();
-  assert(!heap->is_stw_gc_active(), "not reentrant");
-
   IsSTWGCActiveMark mark;
 
-  const bool clear_all_soft_refs =
-    heap->soft_ref_policy()->should_clear_all_soft_refs();
+  ParallelScavengeHeap* heap = ParallelScavengeHeap::heap();
+  clear_all_soft_refs = clear_all_soft_refs
+                     || heap->soft_ref_policy()->should_clear_all_soft_refs();
```

Comment:
In `ParallelScavengeHeap::failed_mem_allocate()`, it does 

```do_full_collection(clear_all_soft_refs = false);
     ....
     do_full_collection(clear_all_soft_refs = true);
     ....
     throws OOM
```

There is possibility that `do_full_collection(clear_all_soft_refs = false)` cycle clears all soft refs due to soft ref policy. In this case, should we bypass the second `do_full_collection()` and throws OOM immediately?

Reply:
That seems a potential optimization. Can we defer it to a followup ticket? (I have some other cleanups in this area, which can affect the design decision here.)

Reply:
Sounds reasonable.